File: frontend/src/api/ai-visibility.ts
import { apiClient } from './client';

// --------------------------------------------------------------------------
// Types
// --------------------------------------------------------------------------
export type AiVisibilityProviderId =
  | 'gemini'
  | 'anthropic'
  | 'openrouter_openai'
  | 'openrouter_anthropic';

export type CompetitorInput = {
  name: string;
  aliases: string[];
  domains: string[];
};

export type PromptInput = {
  text: string;
  theme?: string;
  intent?: string;
};

export type AiVisibilityProject = {
  id: number;
  user_id?: number;
  name: string;
  brand_name: string;
  brand_aliases: string[];
  owned_domains: string[];
  unintended_domains: string[];
  competitors: CompetitorInput[];
  country_code: string;
  language_code: string;
  benchmark_mode: 'consumer_like' | 'controlled_localized' | 'forced_grounded';
  prompts: PromptInput[];
  default_repetitions: number;
  created_at: string;
  updated_at: string;
};

export type AiVisibilityProjectCreate = {
  name: string;
  brand_name: string;
  brand_aliases?: string[];
  owned_domains?: string[];
  unintended_domains?: string[];
  competitors?: CompetitorInput[];
  country_code?: string;
  language_code?: string;
  benchmark_mode?: 'consumer_like' | 'controlled_localized' | 'forced_grounded';
  prompts?: PromptInput[];
  default_repetitions?: number;
};

export type AiVisibilityProjectUpdate = Partial<AiVisibilityProjectCreate>;

export type AiVisibilityRun = {
  id: number;
  project_id: number;
  user_id?: number;
  status: string;
  provider: AiVisibilityProviderId;
  model: string;
  repetitions: number;
  random_seed: string;
  configuration: Record<string, unknown>;
  summary: Record<string, unknown>;
  requested_count: number;
  completed_count: number;
  failed_count: number;
  error_message: string;
  created_at: string;
  updated_at: string;
  completed_at?: string;
};

export type AiVisibilityRunCreate = {
  project_id: number;
  provider?: AiVisibilityProviderId;
  repetitions?: number;
  prompt_indices?: number[];
};

export type AiVisibilityExecution = {
  id: number;
  run_id: number;
  prompt_index: number;
  prompt_text_snapshot: string;
  prompt_theme_snapshot: string;
  prompt_intent_snapshot: string;
  repetition: number;
  randomized_position: number;
  status: string;
  answer_text: string;
  search_used: boolean;
  search_events: Array<{ sequence?: number; query?: string }>;
  citations: Array<Record<string, unknown>>;
  score: Record<string, unknown>;
  request_snapshot: Record<string, unknown>;
  provider_metadata: Record<string, unknown>;
  error_code: string;
  error_message: string;
  latency_ms: number;
};

export type AiVisibilityRunDetail = {
  run: AiVisibilityRun;
  executions: AiVisibilityExecution[];
};

export type AiVisibilityProviderStatus = {
  provider: string;
  label: string;
  surface: string;
  configured: boolean;
  model: string;
  supports_search_fanout: boolean;
  supports_citations: boolean;
};

// --------------------------------------------------------------------------
// API Functions
// --------------------------------------------------------------------------
export async function getProviders() {
  return apiClient.get<AiVisibilityProviderStatus[]>('/api/ai-visibility/providers');
}

export async function getBestAndLessPreset() {
  return apiClient.get<AiVisibilityProjectCreate>('/api/ai-visibility/presets/best-and-less');
}

export async function listProjects(limit = 50) {
  return apiClient.get<AiVisibilityProject[]>(`/api/ai-visibility/projects?limit=${limit}`);
}

export async function getProject(projectId: number) {
  return apiClient.get<AiVisibilityProject>(`/api/ai-visibility/projects/${projectId}`);
}

export async function createProject(payload: AiVisibilityProjectCreate) {
  return apiClient.post<AiVisibilityProject>('/api/ai-visibility/projects', payload);
}

export async function updateProject(projectId: number, payload: AiVisibilityProjectUpdate) {
  return apiClient.patch<AiVisibilityProject>(`/api/ai-visibility/projects/${projectId}`, payload);
}

export async function deleteProject(projectId: number) {
  return apiClient.delete(`/api/ai-visibility/projects/${projectId}`);
}

export async function listRuns(projectId?: number, limit = 50) {
  const params = new URLSearchParams();
  if (projectId !== undefined) params.set('project_id', String(projectId));
  params.set('limit', String(limit));
  const query = params.toString();
  return apiClient.get<AiVisibilityRun[]>(`/api/ai-visibility/runs?${query}`);
}

export async function getRun(runId: number) {
  return apiClient.get<AiVisibilityRunDetail>(`/api/ai-visibility/runs/${runId}`);
}

export async function deleteRun(runId: number) {
  return apiClient.delete(`/api/ai-visibility/runs/${runId}`);
}

export async function cancelRun(runId: number) {
  return apiClient.post<AiVisibilityRun>(`/api/ai-visibility/runs/${runId}/cancel`, {});
}

export async function createRun(payload: AiVisibilityRunCreate) {
  return apiClient.post<AiVisibilityRun>('/api/ai-visibility/runs', payload);
}

export async function getExecution(executionId: number) {
  return apiClient.get<AiVisibilityExecution>(`/api/ai-visibility/executions/${executionId}`);
}

export function getExportCsvUrl(runId: number) {
  return `/api/ai-visibility/runs/${runId}/export.csv`;
}

export function getExportMarkdownUrl(runId: number) {
  return `/api/ai-visibility/runs/${runId}/export.md`;
}

File: backend/app/schemas/ai_visibility.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


# --------------------------------------------------------------------------
# Shared value objects
# --------------------------------------------------------------------------
class CompetitorInput(BaseModel):
    name: str
    aliases: list[str] = Field(default_factory=list)
    domains: list[str] = Field(default_factory=list)


class PromptInput(BaseModel):
    text: str
    theme: str = ""
    intent: str = ""


# --------------------------------------------------------------------------
# Project requests
# --------------------------------------------------------------------------
class AiVisibilityProjectCreate(BaseModel):
    name: str
    brand_name: str
    brand_aliases: list[str] = Field(default_factory=list)
    owned_domains: list[str] = Field(default_factory=list)
    unintended_domains: list[str] = Field(default_factory=list)
    competitors: list[CompetitorInput] = Field(default_factory=list)
    country_code: str = ""
    language_code: str = ""
    benchmark_mode: Literal[
        "consumer_like", "controlled_localized", "forced_grounded"
    ] = "controlled_localized"
    prompts: list[PromptInput] = Field(default_factory=list)
    default_repetitions: int = Field(default=3, ge=1, le=10)


class AiVisibilityProjectUpdate(BaseModel):
    name: str | None = None
    brand_name: str | None = None
    brand_aliases: list[str] | None = None
    owned_domains: list[str] | None = None
    unintended_domains: list[str] | None = None
    competitors: list[CompetitorInput] | None = None
    country_code: str | None = None
    language_code: str | None = None
    benchmark_mode: (
        Literal["consumer_like", "controlled_localized", "forced_grounded"] | None
    ) = None
    prompts: list[PromptInput] | None = None
    default_repetitions: int | None = Field(default=None, ge=1, le=10)


# --------------------------------------------------------------------------
# Responses
# --------------------------------------------------------------------------
class AiVisibilityProjectResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    user_id: int | None = None
    name: str
    brand_name: str
    brand_aliases: list = Field(default_factory=list)
    owned_domains: list = Field(default_factory=list)
    unintended_domains: list = Field(default_factory=list)
    competitors: list = Field(default_factory=list)
    country_code: str
    language_code: str
    benchmark_mode: str
    prompts: list = Field(default_factory=list)
    default_repetitions: int
    created_at: datetime
    updated_at: datetime


class AiVisibilityRunCreate(BaseModel):
    project_id: int
    provider: Literal[
        "gemini", "anthropic", "openrouter_openai", "openrouter_anthropic"
    ] = "gemini"
    repetitions: int | None = Field(default=None, ge=1, le=10)
    prompt_indices: list[int] | None = None


class AiVisibilityRunResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    project_id: int
    user_id: int | None = None
    status: str
    provider: str
    model: str
    repetitions: int
    random_seed: str
    configuration: dict = Field(default_factory=dict)
    summary: dict = Field(default_factory=dict)
    requested_count: int
    completed_count: int
    failed_count: int
    error_message: str
    created_at: datetime
    updated_at: datetime
    completed_at: datetime | None = None


class AiVisibilityExecutionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    run_id: int
    prompt_index: int
    prompt_text_snapshot: str
    prompt_theme_snapshot: str
    prompt_intent_snapshot: str
    repetition: int
    randomized_position: int
    status: str
    answer_text: str
    search_used: bool
    search_events: list = Field(default_factory=list)
    citations: list = Field(default_factory=list)
    score: dict = Field(default_factory=dict)
    request_snapshot: dict = Field(default_factory=dict)
    provider_metadata: dict = Field(default_factory=dict)
    error_code: str
    error_message: str
    latency_ms: int


class AiVisibilityRunDetailResponse(BaseModel):
    run: AiVisibilityRunResponse
    executions: list[AiVisibilityExecutionResponse] = Field(default_factory=list)


class AiVisibilityProviderStatus(BaseModel):
    provider: str
    label: str
    surface: str
    configured: bool
    model: str
    supports_search_fanout: bool
    supports_citations: bool
